src/app/ReadClient.cpp

```diff
@@ -442,78 +441,79 @@ void ReadClient::OnResponseTimeout(Messaging::ExchangeContext * apExchangeContex
+    err = aAttributePath.GetEndpoint(&(aClusterInfo.mEndpointId));
+    VerifyOrReturnError(err == CHIP_NO_ERROR, CHIP_ERROR_IM_MALFORMED_ATTRIBUTE_PATH);
 
-        err = attributePathParser.GetEndpoint(&(clusterInfo.mEndpointId));
-        VerifyOrExit(err == CHIP_NO_ERROR, err = CHIP_ERROR_IM_MALFORMED_ATTRIBUTE_PATH);
+    err = aAttributePath.GetCluster(&(aClusterInfo.mClusterId));
+    VerifyOrReturnError(err == CHIP_NO_ERROR, CHIP_ERROR_IM_MALFORMED_ATTRIBUTE_PATH);
 
-        err = attributePathParser.GetCluster(&(clusterInfo.mClusterId));
-        VerifyOrExit(err == CHIP_NO_ERROR, err = CHIP_ERROR_IM_MALFORMED_ATTRIBUTE_PATH);
+    err = aAttributePath.GetAttribute(&(aClusterInfo.mAttributeId));
+    VerifyOrReturnError(err == CHIP_NO_ERROR, CHIP_ERROR_IM_MALFORMED_ATTRIBUTE_PATH);
 
-        err = attributePathParser.GetAttribute(&(clusterInfo.mFieldId));
-        VerifyOrExit(err == CHIP_NO_ERROR, err = CHIP_ERROR_IM_MALFORMED_ATTRIBUTE_PATH);
+    err = aAttributePath.GetListIndex(&(aClusterInfo.mListIndex));
+    if (CHIP_END_OF_TLV == err)
+    {
+        err = CHIP_NO_ERROR;
+    }
+    VerifyOrReturnError(err == CHIP_NO_ERROR, CHIP_ERROR_IM_MALFORMED_ATTRIBUTE_PATH);
+    VerifyOrReturnError(aClusterInfo.IsValidAttributePath(), CHIP_ERROR_IM_MALFORMED_ATTRIBUTE_PATH);
+    return CHIP_NO_ERROR;
+}
 
-        err = attributePathParser.GetListIndex(&(clusterInfo.mListIndex));
-        if (CHIP_END_OF_TLV == err)
-        {
-            err = CHIP_NO_ERROR;
-        }
-        VerifyOrExit(err == CHIP_NO_ERROR, err = CHIP_ERROR_IM_MALFORMED_ATTRIBUTE_PATH);
+CHIP_ERROR ReadClient::ProcessAttributeReportIBs(TLV::TLVReader & aAttributeReportIBsReader)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    while (CHIP_NO_ERROR == (err = aAttributeReportIBsReader.Next()))
+    {
+        TLV::TLVReader dataReader;
+        AttributeReportIB::Parser report;
+        AttributeDataIB::Parser data;
+        AttributeStatusIB::Parser status;
+        AttributePathIB::Parser path;
+        ClusterInfo clusterInfo;
+        StatusIB statusIB;
 
-        VerifyOrExit(clusterInfo.IsValidAttributePath(), err = CHIP_ERROR_IM_MALFORMED_ATTRIBUTE_PATH);
+        TLV::TLVReader reader = aAttributeReportIBsReader;
+        ReturnErrorOnFailure(report.Init(reader));
 
-        err = element.GetData(&dataReader);
-        if (err == CHIP_END_OF_TLV)
+        err = report.GetAttributeStatus(&status);
+        if (CHIP_NO_ERROR == err)
         {
-            // The spec requires that one of data or status code must exist, thus failure to read data and status code means we
-            // received malformed data from server.
-            SuccessOrExit(err = element.GetStatus(&statusU16));
-            status = static_cast<Protocols::InteractionModel::Status>(statusU16);
+            StatusIB::Parser errorStatus;
+            ReturnErrorOnFailure(status.GetPath(&path));
+            ReturnErrorOnFailure(ProcessAttributePath(path, clusterInfo));
```

Comment:
Followup: `ProcessAttributePath` should just directly write into a ConcreteAttributePath, without the ClusterInfo temporary.